### ISSUE TITLE
[Enhancement] Lock access to a paster for first visitor only

### ISSUE THREAD

**Issue by aldem** [PrivateBin/PrivateBin#134]
Hi,

It would be nice to have an option to lock access to paste to first visitor only - so that person could come back and check again, being sure that nobody else (if get gets the link and encryption key somehow) will be able to peek.

On the other hand, if the first visitor is someone unwanted, it will be obvious that someone did intercept the link (exactly like with "Burn after reading" option).

User IP, User-Agent or special random visitor id set in cookie may be used for this purpose.

**Comment by rugk** [comment]
Good idea.

> User IP

Not really reliable. (think of internet cafès, VPNs/Proxies and so on...)

> User-Agent

Even less reliable and easily fakeable.

> special random visitor id set in cookie

This might be the only possible way. However by default we don't set any persistent cookies in PrivateBin, bit similarly to the language selection option, we could introduce a setting for enabling this feature.

**Comment by aldem** [comment]
Well, sure IP & User-Agent are not reliable enough, though it may be sufficient in cases when receipient has static IP - and combination of both is relatively reliable for those cases, especially when expiration is set. Sure, there is a chance that someone will get/guess those data and attempts to fake it (though it is difficult to fake static IP), but it is small enough.

Option to use any (or all) of three methods would be nice - circumstances are different, and choice is good, and implementation difficulty is not dependent on number of options.

**Comment by elrido** [comment]
I think this is already covered:
* If you want to ensure that only the first visitor of a link can read the message, use a "burn after reading" paste. It will be deleted after the first visit, guaranteeing the subsequent visitors can't read it.
* If the paste needs to be read multiple times by the same visitor, but only one particular visitor, then simply use a password protected paste and share that password only with the single supposed visitor.

Any other solution would require that the server can identify the (first) visitor (via IP, etc.). Such knowledge on the server side would weaken the security of the whole concept.

**Comment by aldem** [comment]
Not really. In second case, if both link *and* password are intercepted by someone, then neither sender or recipient will know that the paste is compromised when the link is accessed by third party (unless they control the server and monitor accesses, of course). Though, access counter could help in this case, but it is relatively unreliable indication as well.

If we salt-hash first visitor's identification (whatever it is - but IP+UA+Cookie is the best), then the knowledge on the server is almost useless (while both IP & UA could be accessed anyway, if someone has access to the server - by itself it provides a lot of information to identify the visitor).

Combining this with "delete" action (to prematurely delete a paste & id cookie before expiration by anyone who is able to decrypt the link) would actually improve security.

But when the server is secure and can be trusted (for instance, corporate server that is used to distribute secure data to employees or customers, like passwords etc), then such knowledge makes no harm at all - the only case when it might be a problem (theoretically) is when maximum anonymity is required.

In any case, if this would be an option, everyone could configure their servers as they see fit - depending on circumstances.

**Comment by marsjaninzmarsa** [comment]
@elrido made a point, this is already covered - if user wants "for me only" paste, it could simply make use of combination of "Burn after reading" and "Clone" option - hashkey stored in browser history will be safer than any combination of IP, useragent or cookies.

**Comment by aldem** [comment]
While technical part (saving a paste) is covered, user experience is definitely not.

To make paste accessible multiple times using "clone", user has to:
1. Click "clone"
2. Set sane options (like expiration)
3. Save the link somewhere in secure place to check it later.

Though as to (3), user may have saved the paste itself in secure place, but this is exactly why I suggest "lock" instead - user may have no time to do so right now.

May be, you always deal with advanced and security minded users, but I often deal with much less advanced users, who often would not follow instruction like "Click clone, save in secure place..." - they click link, see that everything is there, close link and come back later to actually use it.

And hashkey stored in browser history is no more safe - actually, it is a replacement of [ip+ua+]cookie but stored on user side - but this is an advantage only for anonymous users using public services, while PB has many other uses which do not require anonymity (or do not care).  If we use only cookie (which, by the way, may be generated on user side - server only has to remember it), then the security is not sacrificed at all.

Please, do not assume that only users of PB are anonymous paranoid individuals, who are trying hard to hide from everyone - super-puper security measures (quite arguable in this case) are not necessary for less anonymous usage when (for instance) sender is also the server operator and simply distributing data in secure way.

PS: It is not a problem for me to extend PB in this way, I just thought that this feature may be useful for other people as well, thus it makes sense to incorporate it in official code.

**Comment by marsjaninzmarsa** [comment]
Let's think about it:
You're (as hypothetical user) concerned about, well, leaking paste ID and hash, and you want way to secure it, so no one except you, even person having leaked info, cannot access it.

In your scenario, you have constant paste ID and hash. Leaked paste ID and hash.
In my scenario, paste cannot be accessed in any way from old paste ID and hash (our potentially leaked informations).

In your scenario server is your first, last, and only line of defense, so:
1. You must trust the server - admin can if want to just clear your protection setting. I know, ideolo, but motto of this project says about zero knowledge, and I don't think that breaking ideals in favor of (doubtful) user experience can get us anywhere.
2. You must trust the code - in case of security hole your protection can be easily compromised.
3. You must trust anyone in between - PrivateBin don't need HTTPS to provide safe, bulletproof, reliable, private pastebin, but with proposed addition based on cookies, either provider must use HTTPS, or anyone can intercept your private data. Something that cannot be done with hashkey.
In my scenario, everything is still exactly as secure as new paste, everything is still protected on user-side, and every single one protection mechanism of PrivateBin is still used.

To make things clear - I'm not against little UX improvements like this, I'm not against being friendly for non-tech users, I'm only against doing it in stupid, inresponsible way. And no protection at all is still better than false illusion of protection.

You wanna have "Lock" button? OK, let's make it, but **The Right Way**. We can, for example, add new "Allow locking by single user" option in paste creation form, which results in new "Lock" button next to "Clone". After clicking this button, paste can be deleted, content cloned to new one (without form, with default options), and user redirected to new adres. We can solve UX issue on top of the existing mechanics, without need to creating new, ugly and failliable ones.

**Comment by rugk** [comment]
Just some notes about the HTTPS things here, so that things are not got it the wrong way.

> 1. You must trust the server 

You must trust it anyway.

> 2. You must trust the code - in case of security hole your protection can be easily compromised.

Yep, that's actually why you have to trust the server admin & the code & your connection needs to ensure integrity.

> PrivateBin don't need HTTPS to provide safe, bulletproof, reliable, private pastebin, 

That's wrong. To be secure in fact you have to use HTTPS. 
See [the Readme](https://github.com/PrivateBin/PrivateBin#what-it-doesnt-provide) for more information.

Anyway of course I agree that if you use identifying cookies, you **must use** HTTPS as even passive sniffing would otherwise be a privacy exposure. In this case we might do https://github.com/PrivateBin/PrivateBin/issues/64.

**Comment by aldem** [comment]
> You're (as hypothetical user) concerned about, well, leaking paste ID and hash, and you want way to secure it, so no one except you, even person having leaked info, cannot access it.

Almost. I want to be sure that only one person has accessed it - and to know if that person was not my original recipient. And neither sender or recipient care about anonymity - because the sender is the server operator (and the server is reasonably trusted), while the recepient is known to him.

Concerning trust - well, again, it does not matter in my scenario (distributing sensitive data to non-amonymous users) - as the only place where leaks may occur is somewhere outside the server, and this I cannot control anyway.

> I'm only against doing it in stupid, inresponsible way

Sorry, but I don't see where it is irresponsible or stupid. Even assuming that attacker has access to random unique UID he is not able to use it to identify the user. Let's exclude IP+UA from this equation and generate random cookie only - done. If you see something else here which I don't see - please let me know, but I see no way to de-hash IP+UA+Cookie to expose recipient, but even if there is one - this assumes that attacker has access to the server (or is able to decrypt SSL) - in this case he will get IP+UA in any case.

Another option to do it "right way" (but still comfortable to user) is to silently re-encrypt the paste using random cookie (combining it with existing key - this allow to keep the link intact) - thus only right combination of device + user (if password is set) will allow to decrypt the paste. The aim is to minimize user actions in this case - the only thing what he would need is come back to the link (and, probably, enter password). Forcing user to generate and save new link somewhere is the "wrong way" (for UX), and it does not really improve security.

> We can, for example, add new "Allow locking by single user" option in paste creation form, which results in new "Lock" button next to "Clone". 

Well, the idea is to *force* this lock, not to allow the *recipient* to choose - as a sender, I want to be sure that only my recipient on his device (which he used first) will be able to read my paste, and no one else (or, if someone else does it first, recipient will notice this - and let me know - exactly like with "burn after reading").

If we assume that attacker has access to user's device (or can peek behind the user's back) - then we are screwed anyway, nothing can protect content in this case, but access to the server or traffic in transit gives him nothing (well, again, it gives IP+UA - but we cannot prevent it in this case).

And finally, obviously, I did propose to make this an *option* - i.e. server operator (or sender) may choose to use or not use it, thus shifting responsibility to him. As @rugk mentioned in #64, "We should not act as a "moralizer" and let server admins decide things by themself", especially when there is no inherent threat to security.

By the way, coming back to "stupid, inresponsible way" - well, issue #69 clearly demonstrates what is really "irresponsible" - couple of my recipients already asked me "Are you sure this is secure?" when tried to access my links - not counting the fact that this makes impression that the whole thing is unprofessional (while this is not true).
